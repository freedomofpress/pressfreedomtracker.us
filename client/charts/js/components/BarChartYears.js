import * as d3 from 'd3'
import React, { useState } from 'react'
import { groupBy, countBy } from 'lodash'
import { CheckBoxesYear } from './CheckBoxesYear'

const margins = {
  top: 20,
  left: 5,
  right: 40,
  bottom: 1,
}

export function BarChartYears({ data, width, height }) {
  const [selectedYears, setSelectedYears] = useState([])

  const years = countBy(data, (d) => {
    return new Date(d.date).getFullYear()
  })

  const countYears = Object.entries(years).map(([year, count]) => ({
    year: Number(year),
    count,
  }))
  console.log(countYears.map((d) => d.year))

  const xScale = d3
    .scaleLinear()
    .domain(d3.extent(countYears.map((d) => d.year)))
    .range([0 + margins.left, width - margins.right])

  const yScale = d3
    .scaleLinear()
    .domain([0, d3.max(countYears.map((d) => d.count))])
    .range([0, height - margins.bottom - margins.top])

  return (
    <div style={{ flexDirection: 'row' }}>
      <svg width={width} height={height} key={'BarChartYears'}>
        {yScale.ticks(3).map((tick, i) => (
          <g key={i}>
            <line
              x1={margins.left}
              x2={width}
              y1={height - margins.bottom - yScale(tick)}
              y2={height - margins.bottom - yScale(tick)}
              stroke="black"
              strokeWidth={1}
            />
            <text
              x={width}
              y={height - margins.bottom - yScale(tick) - 4}
              textAnchor={'end'}
              fontSize={12}
            >
              {tick}
            </text>
          </g>
        ))}
        {countYears.map((d, i) => (
          <g key={i}>
            <rect
              x={xScale(d.year)}
              y={height - margins.bottom - yScale(d.count)}
              width={10}
              height={yScale(d.count)}
              stroke={'black'}
              fill={selectedYears.includes(d.year) ? '#F2FC67' : 'white'}
              key={d}
            />
          </g>
        ))}
      </svg>
      <div style={{ flexDirection: 'row' }}>
        <CheckBoxesYear
          width={width}
          height={height}
          //options={countYears.map((d) => d.year)}
          options={countYears}
          setSelectedYears={setSelectedYears}
          selectedYears={selectedYears}
        />
      </div>
    </div>
  )
}
