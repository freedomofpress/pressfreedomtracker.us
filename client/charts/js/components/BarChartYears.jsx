import * as d3 from 'd3'
import React from 'react'
import { AnimatedDataset } from 'react-animated-dataset'

const margins = {
	top: 20,
	left: 1,
	right: 40,
	bottom: 1,
}

export default function BarChartYears({ width, height, countYears, selectedYears, onClick }) {
	const internalLeftMargin = 10
	const internalBottomMargin = 20
	const barsWidth = 10

	const xScale = d3
		.scaleLinear()
		.domain(d3.extent(countYears.map((d) => d.year)))
		.range([0 + margins.left + internalLeftMargin, width - margins.right])

	const yScale = d3
		.scaleLinear()
		.domain([0, d3.max(countYears.map((d) => d.count))])
		.range([0, height - margins.bottom - margins.top])
		.nice()

	return (
		<div style={{ flexDirection: 'row', marginBottom: 10 }}>
			<svg
				width={width}
				height={height + internalBottomMargin}
				key={'BarChartYears'}
				style={{ fontFamily: 'var(--font-mono)' }}
			>
				<AnimatedDataset
					dataset={yScale.ticks(3)}
					tag="line"
					init={{
						opacity: 0,
					}}
					attrs={{
						x1: margins.left,
						x2: width,
						y1: (tick) => height - margins.bottom - yScale(tick),
						y2: (tick) => height - margins.bottom - yScale(tick),
						stroke: 'black',
						strokeWidth: (_, i) => (i === 0 ? 3 : 1),
						opacity: 1,
					}}
					keyFn={(d) => d}
					duration={450}
				/>
				<AnimatedDataset
					dataset={yScale.ticks(3)}
					tag="text"
					init={{
						opacity: 0,
					}}
					attrs={{
						x: width,
						y: (tick) => height - margins.bottom - yScale(tick) - 4,
						text: (tick) => tick,
						textAnchor: 'end',
						fontSize: 12,
						opacity: 1,
					}}
					keyFn={(d) => d}
					duration={450}
				/>

				{countYears.map((d, i) => (
					<g key={i}>
						<AnimatedDataset
							dataset={countYears}
							tag="rect"
							init={{
								height: 0,
								y: height,
							}}
							attrs={{
								x: (d) => xScale(d.year),
								y: (d) => height - margins.bottom - yScale(d.count),
								fill: (d) => 'black',
								width: barsWidth,
								stroke: 'black',
								strokeWidth: 2,
								height: (d) => yScale(d.count),
								key: (d) => d,
							}}
							events={{
								onClick: (_, d) => onClick(d),
							}}
							duration={250}
							keyFn={(d) => d.year}
						/>
						<text
							x={xScale(d.year) - internalLeftMargin}
							y={height + internalBottomMargin}
							style={{ fontSize: 12, fontFamily: 'var(--font-base)' }}
						>
							{d.year}
						</text>
					</g>
				))}
				<AnimatedDataset
					dataset={countYears}
					tag="rect"
					init={{
						height: 0,
						y: height,
					}}
					attrs={{
						x: (d) => xScale(d.year),
						y: (d) =>
							selectedYears.includes(d.year) ? height - margins.bottom - yScale(d.count) : height,
						fill: (d) => '#F2FC67',
						width: barsWidth,
						stroke: 'black',
						strokeWidth: 2,
						height: (d) => (selectedYears.includes(d.year) ? yScale(d.count) : 0),
						key: (d, i) => d,
					}}
					events={{
						onClick: (_, d) => onClick(d),
					}}
					duration={250}
					keyFn={(d) => d.year}
				/>
			</svg>
		</div>
	)
}
