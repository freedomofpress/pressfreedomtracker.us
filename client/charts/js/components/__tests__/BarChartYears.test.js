import React from 'react'
import ReactTestRenderer from 'react-test-renderer'
import BarChartYears from '../BarChartYears'

test('renders BarChartYears with mocked data', () => {
	expect(ReactTestRenderer.create(
		<BarChartYears
			countYears={[{"year":2012,"count":2,"numberOfIncidents":0}]}
			x="monthName"
			y="numberOfIncidents"
			selectedYears={[2012]}
			width={480}
			height={500}
		/>
	)).toMatchSnapshot();
});
