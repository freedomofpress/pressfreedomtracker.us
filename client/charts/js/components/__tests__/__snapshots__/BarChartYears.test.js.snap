// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BarChartYears with mocked data 1`] = `
<div
  style={
    {
      "flexDirection": "row",
      "marginBottom": 10,
    }
  }
>
  <svg
    height={520}
    style={
      {
        "fontFamily": "var(--font-mono)",
      }
    }
    width={480}
  >
    <g />
    <g />
    <g>
      <g />
      <text
        style={
          {
            "fontFamily": "var(--font-base)",
            "fontSize": 12,
          }
        }
        x={215.5}
        y={520}
      >
        2012
      </text>
    </g>
    <g />
  </svg>
</div>
`;
